src/store/calendars.js

```diff
@@ -0,0 +1,580 @@
+	 */
+	async getCalendars(context) {
+		let calendars = await client.calendarHomes[0].findAllCalendars()
+			.then(calendars => {
+				return calendars.map(calendar => {
+					return mapDavCollectionToCalendar(calendar)
+				})
+			})
+
+		// remove calendars which don't support tasks
+		calendars = calendars.filter(calendar => calendar.supportsTasks)
+
+		calendars.forEach(calendar => {
+			context.commit('addCalendar', calendar)
+		})
+
+		return calendars
+	},
+
+	/**
+	 * Append a new calendar to array of existing calendars
+	 *
+	 * @param {Object} context the store mutations
+	 * @param {Object} calendar The calendar to append
+	 * @returns {Promise}
+	 */
+	async appendCalendar(context, calendar) {
+		return client.calendarHomes[0].createCalendarCollection(calendar.displayName, calendar.color)
+			.then((response) => {
+				calendar = mapDavCollectionToCalendar(response)
+				context.commit('addCalendar', calendar)
+			})
+			.catch((error) => { throw error })
+	},
+
+	/**
+	 * Delete calendar
+	 * @param {Object} context the store mutations Current context
+	 * @param {Object} calendar the calendar to delete
+	 * @returns {Promise}
+	 */
+	async deleteCalendar(context, calendar) {
+		return calendar.dav.delete()
+			.then((response) => {
+				// delete all the tasks from the store that belong to this calendar
+				Object.values(calendar.tasks)
+					.forEach(task => context.commit('deleteTask', task))
+				// then delete the calendar
+				context.commit('deleteCalendar', calendar)
+			})
+			.catch((error) => { throw error })
+	},
+
+	/**
+	 * Toggle whether a calendar is Enabled
+	 * @param {Object} context the store mutations Current context
+	 * @param {Object} calendar the calendar to toggle
+	 * @returns {Promise}
+	 */
+	async toggleCalendarEnabled(context, calendar) {
+		calendar.dav.enabled = !calendar.dav.enabled
+		return calendar.dav.update()
+			.then((response) => context.commit('toggleCalendarEnabled', calendar))
+			.catch((error) => { throw error })
+	},
+
+	/**
+	 * Change name and color of calendar
+	 * @param {Object} context the store mutations Current context
+	 * @param {Object} data.calendar the calendar to change
+	 * @param {String} data.newName the new name of the calendar
+	 * @param {String} data.newColor the new color of the calendar
+	 * @returns {Promise}
+	 */
+	async changeCalendar(context, { calendar, newName, newColor }) {
+		calendar.dav.displayname = newName
+		calendar.dav.color = newColor
+		return calendar.dav.update()
+			.then((response) => context.commit('renameCalendar', { calendar, newName, newColor }))
+			.catch((error) => { throw error })
+	},
+
+	/**
+	 * Retrieve the tasks of the specified calendar
+	 * and commit the results
+	 *
+	 * @param {Object} context the store mutations
+	 * @param {Object} importDetails = { ics, calendar }
+	 * @returns {Promise}
+	 */
+	async getTasksFromCalendar(context, { calendar }) {
+		return calendar.dav.findByType('VTODO')
```

Comment:
I get a bad request answer from the server for this line.

Parameters:
`<x1:calendar-query xmlns:x1="urn:ietf:params:xml:ns:caldav"><x0:prop xmlns:x0="DAV:"><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x0:displayname/><x0:owner/><x0:resourcetype/><x0:sync-token/><x0:current-user-privilege-set/><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x1:calendar-data/></x0:prop><x1:comp-filter name="VCALENDAR"><x1:comp-filter name="VTODO"/></x1:comp-filter></x1:calendar-query>`

Answer:
`<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\BadRequest</s:exception>
  <s:message>The {urn:ietf:params:xml:ns:caldav}filter element is required for this request</s:message>
</d:error>`

@georgehrke Do you have an idea what might be wrong?

Reply:
I guess something like `<x1:filter> ... </x1:filter>` surrounding the `<x1:comp-filter name="VCALENDAR"><x1:comp-filter name="VTODO"/></x1:comp-filter>` might be missing?

Reply:
Possibly, will look into it tomorrow